File: deletenode.cpp
#include <bits/stdc++.h>
using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};

// Function to delete a node when only that node is given (not the head)
void deleteNode(ListNode* node) {
    ListNode* prev = NULL;
    while (node != NULL && node->next != NULL) {
        node->val = node->next->val;  // Copy value of next node
        prev = node;                  // Keep track of previous node
        node = node->next;            // Move forward
    }
    prev->next = NULL;  // Disconnect the last node
    delete node;        // Free memory
}

int main() {
    // Create linked list: 1 -> 2 -> 3 -> 4
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);

    cout << "Original List: ";
    for (ListNode* temp = head; temp != NULL; temp = temp->next) {
        cout << temp->val << " ";
    }
    cout << endl;

    // Delete node "3" (we pass pointer to node with value 3)
    deleteNode(head->next->next);

    cout << "After deleting node 3: ";
    for (ListNode* temp = head; temp != NULL; temp = temp->next) {
        cout << temp->val << " ";
    }
    cout << endl;

    return 0;
}
/*
DETAILED EXPLANATION:

The problem is: Delete a node from a singly linked list when you are only given
access to that node (not the head). This is a common interview question.

➡ Approach:
1. Instead of removing the given node directly, we copy the data of the *next*
   node into the current node.
2. Then we move forward until the last node, repeating this "copy forward" process.
3. At the end, the last node becomes redundant, so we unlink it and free its memory.

Example:
Original List: 4 -> 5 -> 1 -> 9
Delete node: (5)
Step 1: copy 1 into 5 → 4 -> 1 -> 1 -> 9
Step 2: copy 9 into 1 → 4 -> 1 -> 9 -> 9
Step 3: remove last 9 → 4 -> 1 -> 9

Result: 4 -> 1 -> 9 ✅

⚠ Limitation:
This method cannot delete the last node, since there's no "next node" to copy from.
*/

File: deletenode.java
/*
===============================================================================
Problem: Delete Node in a Linked List (LeetCode 237)
===============================================================================

You are given a node from a singly linked list.
You do NOT have access to the head of the list.

Your task is to delete this node from the linked list.

⚠️ Important Constraints:
- The given node is NOT the last node.
- You cannot traverse backwards.
- You cannot access the head of the list.

Example:
Input Linked List: 4 -> 5 -> 1 -> 9
Node to delete: 5
Output Linked List: 4 -> 1 -> 9

Approach:
- Copy the value of the next node into the current node
- Change the next pointer to skip the next node
- This effectively removes the given node

Time Complexity: O(1)
Space Complexity: O(1)
===============================================================================
*/

import java.util.*;

class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
        this.next = null;
    }
}

public class deletenode{

    // Function to delete the given node (NO head access)
    static void deleteNode(ListNode node) {
        // Copy value of next node into current node
        node.val = node.next.val;

        // Skip the next node
        node.next = node.next.next;
    }

    // Helper function to print linked list
    static void printList(ListNode head) {
        ListNode temp = head;
        while (temp != null) {
            System.out.print(temp.val + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    // Helper function to create linked list from array
    static ListNode createList(int[] arr) {
        ListNode head = new ListNode(arr[0]);
        ListNode curr = head;
        for (int i = 1; i < arr.length; i++) {
            curr.next = new ListNode(arr[i]);
            curr = curr.next;
        }
        return head;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // Input size
        System.out.print("Enter number of nodes: ");
        int n = sc.nextInt();

        int[] arr = new int[n];
        System.out.print("Enter linked list elements: ");
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        // Create linked list
        ListNode head = createList(arr);

        System.out.print("Enter value of node to delete: ");
        int delVal = sc.nextInt();

        // Find node to delete
        ListNode temp = head;
        while (temp != null && temp.val != delVal) {
            temp = temp.next;
        }

        if (temp == null || temp.next == null) {
            System.out.println("Invalid node (cannot delete last node)");
            return;
        }

        // Delete the node
        deleteNode(temp);

        System.out.print("Linked List after deletion: ");
        printList(head);
    }
}

// class Solution {
//     public void deleteNode(ListNode node) {
//         node.val = node.next.val;      // Copy next node's value
//         node.next = node.next.next;    // Skip the next node
//     }
// }